pkg/R/utils.R

```diff
@@ -16,51 +16,41 @@ convertJListToRList <- function(jList, flatten, logicalUpperBound = NULL, serial
   results <- if (arrSize > 0) {
     lapply(0:(arrSize - 1),
            function(index) {
-             jElem <- .jcall(jList,
-                             "Ljava/lang/Object;",
-                             "get",
-                             as.integer(index))
+             obj <- callJMethod(jList, "get", as.integer(index))
 
              # Assume it is either an R object or a Java obj ref.
-             obj <- .jsimplify(jElem)
-
-             if (inherits(obj, "jobjRef") && .jinstanceof(obj, "[B")) {
-               # RDD[Array[Byte]]. `obj` is a whole partition.
-
-               rRaw <- .jevalArray(.jcastToArray(jElem))
-               res <- unserialize(rRaw)
-
-               # For serialized datasets, `obj` (and `rRaw`) here corresponds to
-               # one whole partition dense-packed together. We deserialize the
-               # whole partition first, then cap the number of elements to be returned.
-
-               # TODO: is it possible to distinguish element boundary so that we can
-               # unserialize only what we need?
-               if (!is.null(logicalUpperBound)) {
+             if (inherits(obj, "jobj")) {
+               if (isInstanceOf(obj, "scala.Tuple2")) {
+                 # JavaPairRDD[Array[Byte], Array[Byte]].
+
+                 keyBytes = callJMethod(obj, "_1")
+                 valBytes = callJMethod(obj, "_2")
+                 res <- list(unserialize(keyBytes),
+                             unserialize(valBytes))
+               } else {
+                 stop(paste("utils.R: convertJListToRList only supports",
+                            "RDD[Array[Byte]] and",
+                            "JavaPairRDD[Array[Byte], Array[Byte]] for now"))
+               }
+             } else {
+               if (inherits(obj, "raw")) {
+                 # RDD[Array[Byte]]. `obj` is a whole partition.
+                 res <- unserialize(obj)
+                 # For serialized datasets, `obj` (and `rRaw`) here corresponds to
+                 # one whole partition dense-packed together. We deserialize the
+                 # whole partition first, then cap the number of elements to be returned.
+
+                 # TODO: is it possible to distinguish element boundary so that we can
+                 # unserialize only what we need?
+                 if (!is.null(logicalUpperBound)) {
                    res <- head(res, n = logicalUpperBound)
+                 }
+               } else {
+                 # jElem is of a primitive Java type, is simplified to R's
```

Comment:
Yeah so this is the case that gets called when we have a `JavaRDD[String]` that gets `collect` called on it. In this case the java result is `ArrayList[String]` and when we do a `get` on the ArrayList, our deserializer automatically converts `String` to `character`